### PR TITLE
[WIP] Add workflow step parameters to workflow editor

### PR DESCRIPTION
**Pull request by mvdbeek**
This should allow specifying text, integer, float, boolean and color Parameters that can be re-used across workflow nodes. This build on https://github.com/galaxyproject/galaxy/pull/1306/commits/45962f32344a00162f5d04b94096c8a570f2adf1

### CODE REVIEW

**Comment by mvdbeek** [comment]
A good fraction of the UI is there and seems to work OK-ish, but the actual non-data parameter replacement only works for new-style workflow API submissions with `inputs`. The workflow run form still uses a different format for this. I'll see if it is sufficient to just move `input_parameter` to inputs, but IIRC there was a longer-term plan to unify UI and API. That isn't happening anytime soon, right ?

File: lib/galaxy/workflow/modules.py [hunk 1]
```diff
@@ -1244,27 +1256,19 @@ def load_module_sections(trans):
...
+            },
+            {
+                "name": "parameter_input",
+                "title": "Parameter Input",
+                "description": "Simple inputs used for workflow logic"
+            },
         ],
     }
 
-    if trans.app.config.enable_beta_workflow_modules:
```
Comment (jmchilton) [inline lib/galaxy/workflow/modules.py:1250]: My two cents is that I'd keep the option and just promote parameter_input to an un-beta parameter.

**Comment by jmchilton** [comment]
I'll play with this and see if I can figure out an opinion on the API unification and fix the tests.

**Comment by jxtx** [comment]
Oooooh. Screenshots?

**Comment by mvdbeek** [comment]
Nothing fancy yet:
screen shot 2018-10-11 at 17 50 32
screen shot 2018-10-11 at 17 51 42

We should separate the data and non-data inputs in the workflow node (a bar, and maybe different color noodle), and the workflow inputs should move to the top where we also render the old workflow replacement parameters. And I'm sure there are plenty of bugs in there, just trying to figure this out.

**Comment by jmchilton** [comment]
> The workflow run form still uses a different format for this. I'll see if it is sufficient to just move input_parameter to inputs, but IIRC there was a longer-term plan to unify UI and API. That isn't happening anytime soon, right ?

:( indeed it does. I'll start with the input parameters and see if I can separate them out, after that if I can separate out the data parameters so we are tracking them with database integrity using proper tracking with  ``WorkflowRequestToInputDatasetAssociation`` / ``WorkflowRequestToInputDatasetCollectionAssociation`` the way standardized API calls with generate.

**Comment by jmchilton** [comment]
Working on ripping parameter inputs from the raw param map produced by the tool form and I encountered another problem - it looks like this branch breaks the parameter_inputs in some way. In particular the following test case that works in dev:

```
./run_tests.sh -api test/api/test_workflows.py::WorkflowsApiTestCase::test_run_with_text_connection
```

No longer works, the key line in the logs I think is seeing that there is an unmatched input connection.

```
galaxy.workflow.modules WARNING 2018-10-12 10:31:42,351 Failed to use input connections for inputs [set([u'seed_source|seed'])]
```

I'm going to see if I can figure out what caused this - none of the backend changes are obvious problems at first glance to me.

File: lib/galaxy/model/__init__.py [hunk 2]
```diff
@@ -4179,14 +4179,10 @@ def __init__(self):
...
-    def set_non_data_connection(self):
-        self.output_name = WorkflowStepConnection.NON_DATA_CONNECTION
-        self.input_name = WorkflowStepConnection.NON_DATA_CONNECTION
-
     @property
     def non_data_connection(self):
-        return (self.output_name == WorkflowStepConnection.NON_DATA_CONNECTION and
-                self.input_name == WorkflowStepConnection.NON_DATA_CONNECTION)
+        return (self.output_name == self.input_name == WorkflowStepConnection.NON_DATA_CONNECTION or
+                self.output_step and self.output_step.type == 'parameter_input')
```
Comment (jmchilton) [inline lib/galaxy/model/__init__.py]: Reverting this diff fixes the test case. I guess there aren't docs for this - but I don't like this change. The parameter_input is transferring data to the step - it isn't an implicit connection so this should return False IMO. I see why you made the change here - for when building the editor dict:

```diff
                 # FIXME: this removes connection without displaying a message currently!
+                input_connections = [conn for conn in input_connections if conn.input_name in data_input_names]
-                input_connections = [conn for conn in input_connections if conn.input_name in data_input_names or conn.non_data_connection]
```

Maybe we should bite the bullet and just not do that filtering at all - push all the connections to the editor? This way the editor can tell the user if there are invalid connections found and we can eliminate that FIXME?

Comment (mvdbeek) [reply lib/galaxy/model/__init__.py]: Sounds good to me!

File: client/galaxy/scripts/mvc/workflow/workflow-view-terminals.js [hunk 3]
```diff
@@ -253,10 +264,10 @@ var OutputTerminalView = BaseOutputTerminalView.extend({
     terminalMappingClass: Terminals.TerminalMapping,
     terminalMappingViewClass: TerminalMappingView,
     terminalForOutput: function(output) {
-        var type = output.extensions;
-        var terminal = new Terminals.OutputTerminal({
+        var type = output.type;
+        var terminal = new Terminals.OutputParameterTerminal({
             element: this.el,
-            datatypes: type
```
Comment (jmchilton) [inline client/galaxy/scripts/mvc/workflow/workflow-view-terminals.js]: Should these changes be reverted - given the class below?

Comment (mvdbeek) [reply client/galaxy/scripts/mvc/workflow/workflow-view-terminals.js]: Oh yeah, sure, should have cleaned this up before.

**Comment by jmchilton** [comment]
@mvdbeek any chance I can take this over this week and rebase #6776 underneath it? I'd like to add some Selenium tests and see if I can implement default values, but I can wait on you and limit the scope at how it stands now if you'd like.

**Comment by mvdbeek** [comment]
Sure, go ahead, happy to get this rolling!

**Comment by jmchilton** [comment]
@mvdbeek My new change is still creating more issues then solving and the GUI is terrible - but this is an outline of separating RuntimeValue into RuntimeValue and ConnectedValue (https://github.com/jmchilton/galaxy/commit/15e667cdf9b566b7167d53954f8da9d57b54d10f).

I guess my next step is to try to make data and non-data parameters more similar in terms of the tool state and module interface. The data parameters should have ConnectedValue as their tool state value and then we can uniformly have inputs with ConnectedValues appear with connections and eliminate the distinction between parameters and inputs. This will be good from an interface perspective but is also required for CWL where a parameter that can take an integer or a File for instance is totally possible. I'll probably merge https://github.com/galaxyproject/galaxy/pull/6911 into that to provide that consistent interface on the backend, I'll rework the frontend a bit - I think you have the right classes I just want to change how their instantiated a bit I think.

File: client/galaxy/scripts/mvc/workflow/workflow-view-terminals.js (hunk 3 above)
Comment (jmchilton) [reply client/galaxy/scripts/mvc/workflow/workflow-view-terminals.js]: Going to remove this via rebase on this commit and force push to your branch - hope that is okay.

Comment (mvdbeek) [reply client/galaxy/scripts/mvc/workflow/workflow-view-terminals.js]: sure

**Pull request closed without merging**